crates/fixedpointmath/src/fixed_point.rs

```diff
@@ -0,0 +1,399 @@
+use std::fmt::{self, Debug};
+
+use ethers::types::{I256, U256};
+use eyre::{bail, eyre, Result};
+
+use crate::{sign::FixedPointSign, utils::u256_from_str, value::FixedPointValue};
+
+/// A generic fixed point wrapper around the `U256` type from ethers-rs.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct FixedPoint<T: FixedPointValue> {
+    raw: T,
+    decimals: u8,
+}
+
+impl<T: FixedPointValue> FixedPoint<T> {
+    pub const MIN: Self = Self {
+        raw: T::MIN,
+        decimals: T::MAX_DECIMALS,
+    };
+
+    pub const MAX: Self = Self {
+        raw: T::MAX,
+        decimals: T::MAX_DECIMALS,
+    };
+
+    // Constructors //
+
+    pub fn new<V: Into<T>>(value: V) -> Self {
+        Self {
+            raw: value.into(),
+            // TODO: Add support for variable decimals.
+            decimals: T::MAX_DECIMALS,
+        }
+    }
+
+    pub fn try_from<V: TryInto<T> + Debug>(value: V) -> Result<Self> {
+        // Convert the value to a Debug string before moving it incase the
+        // conversion fails.
+        let value_debug = format!("{:?}", value);
+        let value = value.try_into().map_err(|_| {
+            eyre!(
+                r#"Failed to convert value to underlying FixedPointValue type:
+    value: {value_debug}
+    Underlying range: {min:?} to {max:?}
+"#,
+                min = Self::MIN,
+                max = Self::MAX,
+            )
+        })?;
+
+        Ok(Self::new(value))
+    }
+
+    pub fn from_sign_and_abs(sign: FixedPointSign, abs: U256) -> Result<Self> {
+        Ok(match sign {
+            FixedPointSign::Positive => Self::new(T::from_u256(abs)?),
+            FixedPointSign::Negative => {
+                if abs == T::MIN.unsigned_abs() {
+                    // Avoid overflow
+                    Self::MIN
+                } else {
+                    let raw = T::from_u256(abs)?.flip_sign();
+                    Self::new(raw)
+                }
+            }
+        })
+    }
+
+    pub fn from_dec_str(s: &str) -> Result<Self> {
```

Comment:
Ah, yeah. I had originally planned to support adding signs to unsigned integers (e.g., a negative `U256`), but:

- Keeping the raw value sign in sync with the outer type gets messy and inneficient.
- A negative `U256` that's too low to be converted to a `I256` is kinda hard to work with since the `raw` value can't be represented using the underlying type.
- Lastly, two's compliment is already a great and efficient system for representing signed integers; no need for me to get *too* clever.

Instead, the lib now inherits the sign from the underlying type. If you plan to use negative numbers, you'd use an `i128` or `I256`. This way the behavior is predictable and calling `raw()` will return the actual underlying value, sign and all.

Reply:
Ah, yeah. I had originally planned to support adding signs to unsigned integers (e.g., a negative `U256`), but:

- Keeping the raw value sign in sync with the outer type gets messy and inneficient.
- A negative `U256` that's too low to be converted to a `I256` is kinda hard to work with since the `raw` value can't be represented using the underlying type.
- Lastly, two's compliment is already a great and efficient system for representing signed integers; no need for me to get *too* clever.

Instead, the lib now inherits the sign from the underlying type. If you plan to use negative numbers, you'd use an `i128` or `I256`. This way the behavior is predictable and calling `raw()` will return the actual underlying value, sign and all.